source/slang/slang-capability.cpp

```diff
@@ -134,1201 +140,954 @@ bool isCapabilityDerivedFrom(CapabilityAtom atom, CapabilityAtom base)
-        }
+    return true;
+}
+
+/// Join a compatable target set from `this` with `CapabilityTargetSet other`.
+/// Return false when `other` is fully incompatable.
+/// incompatability is when one of 2 senarios are true:
+/// 1. `this->target` is not a supported target by `other.shaderStageSets`
+/// 2. `this` has completly disjoint shader stages from other.
+bool CapabilityTargetSet::tryJoin(const CapabilityTargetSets& other)
+{
+    const CapabilityTargetSet* otherTargetSet = other.tryGetValue(this->target);
+    if (otherTargetSet == nullptr)
+        return false;
+
+    List<CapabilityAtom> destroySet;
+    destroySet.reserve(this->shaderStageSets.getCount());
+    for (auto& shaderStageSet : this->shaderStageSets)
+    {
+        if (!shaderStageSet.second.tryJoin(*otherTargetSet))
+            destroySet.add(shaderStageSet.first);
     }
+    if (destroySet.getCount() == Slang::Index(this->shaderStageSets.getCount()))
+        return false;
 
-    return false;
+    for (const auto& i : destroySet)
+        this->shaderStageSets.remove(i);
+
+    return true;
 }
 
-bool CapabilityConjunctionSet::implies(CapabilityConjunctionSet const& that) const
+void CapabilitySet::join(const CapabilitySet& other)
 {
-    // One capability set implies another if it is a super-set
-    // of the other one. Think of it this way: if your target
-    // supports features {X, Y, Z}, then that implies it also
-    // supports features {X,Z}.
-    //
-    // Because both `this` and `that` have expanded lists
-    // of all the capability atoms they imply *and* those
-    // lists are sorted, we can simply walk through the
-    // lists in tandem and see if there are any entries
-    // in `that` which are not present in `this.
+    if (isEmpty() || other.isInvalid())
+    {
+        *this = other;
+        return;
+    }
+    if (isInvalid())
+        return;
+    if (other.isEmpty())
+        return;
 
-    Index thisCount = this->m_expandedAtoms.getCount();
-    Index thatCount = that.m_expandedAtoms.getCount();
+    bool isEmptyBefore = this->m_targetSets.getCount() == 0;
+    List<CapabilityAtom> destroySet;
+    destroySet.reserve(this->m_targetSets.getCount());
+    for (auto& thisTargetSet : this->m_targetSets)
+    {
+        if (!thisTargetSet.second.tryJoin(other.m_targetSets))
+        {
+            destroySet.add(thisTargetSet.first);
+        }
+    }
+    for (const auto& i : destroySet)
+    {
+        this->m_targetSets.remove(i);
+    }
+    // join made a invalid CapabilitySet
+    if (!isEmptyBefore && this->m_targetSets.getCount() == 0)
+        this->m_targetSets[CapabilityAtom::Invalid].target = CapabilityAtom::Invalid;
+}
 
-    // We cannot possibly have `this` contain all the atoms
-    // in `that` if the latter is has more atoms.
-    //
-    if(thatCount > thisCount)
-        return false;
+static uint32_t _calcAtomListDifferenceScore(List<CapabilityAtom> const& thisList, List<CapabilityAtom> const& thatList)
```

Comment:
We need the _info of a capabilityAtom. This means we only can do the processing on a `List<CapabilityAtom>`.

Reply:
I see your point. I will do that

Reply:
oh, I see why I did not:
```
                auto thisDiffScore = _calcAtomListDifferenceScore(thisSetElements, shaderStageSetWeNeedElements);
                auto thatDiffScore = _calcAtomListDifferenceScore(thisSetElements, shaderStageSetWeNeedElements);
```
to avoid calling `auto shaderStageSetWeNeedElements = shaderStageSetWeNeed.getElements<CapabilityAtom>();` twice inside `_calcAtomListDifferenceScore`

Reply:
We should be able to use UIntSet just like List<CapabilityAtom> if it supports iterating, right?